### PR TITLE
CUDA support in the CSR layout:  CUDA addmm

### PR DESCRIPTION
**Pull request by aocsa**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57403 CUDA support in the CSR layout:  CUDA addmm**
* #57275 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr
* #57274 CUDA support in the CSR layout: constructors

This PR add  CUDA support for  matvec/addmm ops for CSR Layout.   

Differential Revision: [D28161161](https://our.internmc.facebook.com/intern/diff/D28161161)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28161161).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28161161).

**Comment by aocsa** [comment]
I rebased this PR and polished a few things. As a gentle reminder, this PR is still waiting for a final review before merging, and I won't be able to work on this after this month. cc @pearu, @ezyang

**Review by malfet — APPROVED** [review]

File: test/test_sparse_csr.py
Comment (pearu) [inline test/test_sparse_csr.py]: Test this for CPU as well?

File: test/test_sparse_csr.py
Comment (pearu) [inline test/test_sparse_csr.py]: For completeness, CSR tensors should be tested for both int32 and int64 indices.

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (pearu) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:67]: Start `if (nnz > 0)` block from here.

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (pearu) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:105]: The implementation of addmm seems to be incomplete: if `nnz==0`, currently, the result will be `0`, but it should be `beta * r_`.

**Review by pearu — CHANGES_REQUESTED** [review]
Thanks, @aocsa !

The implementation of addmm seems to be incomplete, IIUC: when `nnz == 0`, the result is all zeros but it ought to be `beta * input`.

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py]: Yes, This PR was waiting the merge of PR https://github.com/pytorch/pytorch/pull/58768

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28161161).

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:102]: Thanks Pearu for the comment. Unfortunately, cuSparse/csrmm2 doesn't support int64 indices. 

https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu#L200

I will remove these warnings to be consistent with .
https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu#L62

Comment (pearu) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:102]: > Thanks Pearu for the comment. Unfortunately, cuSparse/csrmm2 doesn't support int64 indices.

See https://github.com/pytorch/pytorch/blob/e4be80c1b82a907fd1c799624af5987ba1e843b3/aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu#L125-L126

IIUC, the task boils down to generalizing `template<typename T> void _csrmm2...` to `template<typename T, typename I>
void _csrmm2` where `I` is index type.

If this makes sense, add `TODO` comment for follow-up work.

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:81]: Seems ok. Moreover I added a test for transposed input tensors

**Comment by aocsa** [comment]
> Thanks, @aocsa !
> 
> The implementation of addmm seems to be incomplete, IIUC: when `nnz == 0`, the result is all zeros but it ought to be `beta * input`.

Thank u @pearu.  I addressed the feedback comments and added new test cases.  Let me know your thoughts.

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (pearu) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:80]: How does this seem ok? For instance, it does not match the same pattern used above: `if(dense.stride(0) == 1 && dense.stride(1) == dense.size(0))`..

So, let's analyze this: `dense` is a 2D tensor and if its last stride (`stride(1)`) is 1, then the first stride must be `size(1) * s` where `s` is some positive integer. Here, `dense` must be contiguous, hence, `s == 1` and `stride(0) == size(1)`, in general. As a result, the code block under `if(dense.stride(1) == 1 && dense.stride(0) != dense.size(1))` is never reached.

The test cases most likely use the `else` code block, so tests will not discover this bug.

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu]: Sure, make sense.

Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu]: I will create an issue for this too as it could affect other things.  
NVIDIA docs mention that CUSPARSE_INDEX_64I can be used but for CUSPARSE_SPMM_CSR_ALG2/CUSPARSE_SPMM_CSR_ALG3 currently is it uses CUSPARSE_CSRMM_ALG1
https://docs.nvidia.com/cuda/cusparse/index.html

> cusparseSpMM supports the following index type for representing the sparse matrix matA:
> 32-bit indices (CUSPARSE_INDEX_32I)
> 64-bit indices (CUSPARSE_INDEX_64I) only with CUSPARSE_SPMM_COO_ALG4, CUSPARSE_SPMM_CSR_ALG2, and CUSPARSE_SPMM_CSR_ALG3 algorithms

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:80]: Sorry I misunderstood the comment. Now 2nd condition is `else if(dense.stride(1) == 1 && dense.stride(0) == dense.size(1)) `

File: aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu
Comment (aocsa) [reply aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu]: Mentioned here: https://github.com/pytorch/pytorch/issues/58899

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (pearu) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:67]: With `if (nnz == 0)` used in the caller functions, this `if` condition can now be eliminated.
Use `TORCH_INTERNAL_ASSERT(nnz > 0)` instead.

File: aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu
Comment (pearu) [inline aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:71]: Not sure what this TODO is about. What the code does below, is just ensuring that `r__` is Fortran contiguous.
Perhaps add a comment explaining why this is preferred?

**Review by pearu — APPROVED** [review]
I have a couple of nit comments. Otherwise, LGTM. Thanks, @aocsa!

**Comment by aocsa** [comment]
> I have a couple of nit comments. Otherwise, LGTM. Thanks, @aocsa!

Thanks @pearu.

**Comment by aocsa** [comment]
I think this stack  PR needs to re-import PR to be merged

**Comment by aocsa** [comment]
Closing this PR too :(. Stack PR was poisoned after first  PR in the stack was reverted. New PR is here:
https://github.com/pytorch/pytorch/pull/59004

**Comment by bhosmer** [comment]
new new PR: new new PR: #59012 (thanks for restacking @aocsa)

**Pull request closed without merging**
